src/main/java/de/tum/cit/aet/artemis/iris/service/session/AbstractIrisChatSessionService.java

```diff
@@ -14,6 +16,8 @@ public abstract class AbstractIrisChatSessionService<S extends IrisChatSession>
 
     private final ObjectMapper objectMapper;
 
+    protected final HashMap<String, LLMTokenUsageTrace> traces = new HashMap<>();
```

Comment:
I don't think this is the right place for this attribute - it is only ever used in a single subclass.